internal/pkg/vm/state/state.go

```diff
@@ -0,0 +1,204 @@
+type Root = cid.Cid
+
+type actorKey = address.Address
+
+// State is the VM state manager.
+type State struct {
+	store    cbor.IpldStore
+	dirty    bool
+	root     Root
+	rootNode *hamt.Node
+}
+
+// NewState creates a new VM state.
+func NewState(store cbor.IpldStore) *State {
+	// Review: we should probably start the tree as dirty given that the current root does not match the state it will persist
+	return newState(store, cid.Undef, hamt.NewNode(store, hamt.UseTreeBitWidth(TreeBitWidth)))
+}
+
+// LoadState creates a new VMStorage.
+func LoadState(ctx context.Context, store cbor.IpldStore, root Root) (*State, error) {
+	// Review: do we want to special case root=cid.Undef to be a new state?
+	rootNode, err := hamt.LoadNode(ctx, store, root, hamt.UseTreeBitWidth(TreeBitWidth))
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load node for %s", root)
+	}
+
+	return newState(store, root, rootNode), nil
+}
+
+func newState(store cbor.IpldStore, root Root, rootNode *hamt.Node) *State {
+	return &State{
+		store:    store,
+		dirty:    false,
+		root:     root,
+		rootNode: rootNode,
+	}
+}
+
+// GetActor retrieves an actor by their key.
+// If the actor is not found it will return false and no error.
+// If there are any IO or decoding errors, it will return false and the error.
+func (st *State) GetActor(ctx context.Context, key actorKey) (*actor.Actor, bool, error) {
+	actorBytes, err := st.rootNode.FindRaw(ctx, string(key.Bytes()))
+	if err == hamt.ErrNotFound {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	var act actor.Actor
+	err = encoding.Decode(actorBytes, &act)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return &act, true, nil
+}
+
+// SetActor sets the the actor to the given value wheter it previously existed or not.
+//
+// This method will not check if the actor previuously existed, it will blindly overwrite it.
+func (st *State) SetActor(ctx context.Context, key actorKey, a *actor.Actor) error {
+	actBytes, err := encoding.Encode(a)
+	if err != nil {
+		return err
+	}
+	if err := st.rootNode.SetRaw(ctx, string(key.Bytes()), actBytes); err != nil {
+		return errors.Wrap(err, "setting actor in state tree failed")
+	}
+	st.dirty = true
+	return nil
+}
+
+// DeleteActor remove the actor from the storage.
+//
+// This method will NOT return an error if the actor was not found.
+func (st *State) DeleteActor(ctx context.Context, key actorKey) error {
+	err := st.rootNode.Delete(ctx, string(key.Bytes()))
+	st.dirty = true
+	if err == hamt.ErrNotFound {
+		return nil
+	}
+	return err
+}
+
+// Commit will flush the state tree into the backing store.
+// The new root is returned.
+func (st *State) Commit(ctx context.Context) (Root, error) {
+	if err := st.rootNode.Flush(ctx); err != nil {
+		return cid.Undef, err
+	}
+
+	// Review: is this put needed because it has not been flushed or because flush does not return the root cid?
```

Comment:
I believe the root node is not `Put` in the store on `Flush`, only children are (so Flush doesn't yet have the CID).